### ISSUE TITLE
 Echo 11 Thunderbolt 4 ECHO-DK11-T4 heavily impacting computer's performance 

### ISSUE THREAD

**Issue by wiktormowinski** [Dasharo/dasharo-issues#980]
### Component

other

### Device

NovaCustom V54 14th Gen

### Dasharo version

v0.9.0 

### Dasharo Tools Suite version

-

### Test case ID

-

### Brief summary

There is a problem to reproduce client's evironment to get the same result as them

### How reproducible

not sure

### How to reproduce

that's the issue- we don't know

### Expected behavior

The platform's performance is eavily impacted by dock

### Actual behavior

Everything works smoothly.

### Additional context

The client complained about the performance of their computer while using Echo 11 Thunderbolt 4 ECHO-DK11-T4.
However, while running stress tests to reproduce such behaviour, the machines CPU clock and power remained relatively the same regardless if the dock was plugged or not. Therefore further investigation of this issue needed.

### Solutions you've tried

current approach aimed at measuring CPU's performance, perhaps the problem lies somewhere else entirely

**Comment by zirblazer** [comment]
I recall some guy on chat that also had performance issues when plugging a TB4 dock (Not the same one, but maybe they share a reference design/chips). IGP performance halved by just having the dock plugged in.

**Comment by PerAstraAdDeum** [comment]
> I recall some guy on chat that also had performance issues when plugging a TB4 dock (Not the same one, but maybe they share a reference design/chips). IGP performance in a graphics benchmark halved by just having the dock plugged in.

It was I who reported this issue on Matrix. Here's what I know so far:

On my own TB4 docking station ([Icy Box  IB-DK8801-TB4 Thunderbolt™ 4 Dock with PD 100 W](https://icybox.de/product/dockingstation/IB-DK8801-TB4)), having the notebook connected and powered results in a performance drop of roughly 50% compared to the barrel-jack power connector.
The Icy Box  IB-DK8801-TB4 is a rebrand of the [Goodway DBD1330 Thunderbolt™ 4 / USB4 Dock Pro](https://www.goodway.com.tw/en/product/detail/DBD1330), which itself is based on Goshen Ridge Thunderbolt Controller afaik.

I believe the issue in question is related to power delivery. When having both the dock as well as the barrel-jack power plug connected, the notebooks performance is as expected.

**Comment by wiktormowinski** [comment]
https://pad.3mdeb.com/sheet/#/2/sheet/view/3qtlwsbhPvu3fo1ZgaTfwMUC2ia0eWj2nY21+lKTky8/

**Comment by zirblazer** [comment]
> Results of mentioned tests are included here: https://pad.3mdeb.com/sheet/#/2/sheet/view/3qtlwsbhPvu3fo1ZgaTfwMUC2ia0eWj2nY21+lKTky8/

Note that you are including CPU results only. The benchmarks that were catastrophic (50% difference) were GPU related. You most likely want @PerAstraAdDeum to post his results here and what was his test environment.

**Comment by wiktormowinski** [comment]
valid point, here are gpu logs: 
[gpu-dock-stress-test.log](https://github.com/user-attachments/files/16943781/gpu-dock-stress-test.log)
[gpu-stress-test.log](https://github.com/user-attachments/files/16943782/gpu-stress-test.log)

**Comment by zirblazer** [comment]
Well, those do seem within error margin. So I suppose we need more info from the affected users setups.

**Comment by wiktormowinski** [comment]
yeah :c

**Comment by PerAstraAdDeum** [comment]
After two more months I did another benchmark. Here's the result (last line):

![Screenshot_20241109_131156](https://github.com/user-attachments/assets/23399b1d-6387-4656-b922-e7790eaaa994)

As you can see, the error is persistent. There's also a performance regress for the iGPU of about ten percent, that also occurs when powered by the barrel jack (4820 -> 4423, that's  -8,24%). Don't know the cause of that yet.

On the bright side, it seems like only the iGPU is affected. Benchmarks with my eGPU are consistent between barrel-jack-powered and TB4-powered.

**Comment by PerAstraAdDeum** [comment]
Well, seems like we're back to where we started:

![image](https://github.com/user-attachments/assets/7276f1a6-6c89-49cd-9e3a-bfe6d6836d9f)

Now this is a benchmark done just a few minutes ago.

Coincidentally, I've been helping a dev testing some new patches for Mangohud, and I marveled at the performance of the Intel Arc iGPU:

![image](https://github.com/user-attachments/assets/6c357b32-269c-4cf8-8912-bf8cd1ad14ea)

When I start the game now, my framerate is considerably lower:

![image](https://github.com/user-attachments/assets/f1c34cda-3e68-441b-b4d7-b28070887da4)

As you can see by the time shown in the HUD, there's only about 90 minutes between these runs. So whatever did impact the performance must have happened in these 90 minutes.
And indeed, something happened: the TB4 connection was interrupted, and I had to re-plug the cable.
I'm pretty sure now that this is what did halve the performance. Sadly, it prevailed a reboot. I'll try a cold boot next and see if the issue prevails.

**Comment by PerAstraAdDeum** [comment]
A cold boot didn't solve the issue.

I was however able to trick the notebook like this:

1. Plug both TB4 as well as barrel-jack charger into the notebook.
2. Run benchmark/graphics-demanding application.
3. Pull barrel-jack charger.

Surprisingly I'm now back to "full-power-mode", if you will:

![image](https://github.com/user-attachments/assets/e07c2661-3c21-4d43-b762-e95d91b841f0)

This is obviously just a hack and no viable solution.

**Comment by miczyg1** [comment]
@mkopec isn't that because EC just cuts the power by cropping PL4 to the max USB-PD power? When barrel jack is plugged, then PL4 is higher and allows to reach >4000 score. Just a hunch

**Comment by PerAstraAdDeum** [comment]
> @mkopec isn't that because EC just cuts the power by cropping PL4 to the max USB-PD power? When barrel jack is plugged, then PL4 is higher and allows to reach >4000 score. Just a hunch

Is there any way of testing this?

**Comment by mkopec** [comment]
> @mkopec isn't that because EC just cuts the power by cropping PL4 to the max USB-PD power? When barrel jack is plugged, then PL4 is higher and allows to reach >4000 score. Just a hunch

That's what I thought too.

> Is there any way of testing this?

Check `Limit reasons` in ThrottleStop (on Windows).

The dock itself can deliver 90W which is as good as the barrel jack charger. Question is, does it actually deliver that much? That can depend on the USB-C cable. Some cables are rated for 60W max, for example, so if the cable was changed at some point, that could be the culprit. @PerAstraAdDeum Can you check the maximum power draw in throttlestop while on dock alone and while on barrel jack charger?

**Comment by PerAstraAdDeum** [comment]
Hey, thanks for the swift reply!

Sadly I'm not running Windows. Is there a way to read those values in Linux? (Arch to be precise)

Also, about the cable. A few months ago I've changed the dock's TB4 cable for this one here: [Cable Matters [Intel Certified] 40Gbps Braided Active Thunderbolt 4 Cable 2 m with 100W Charging Power](https://www.amazon.de/-/en/Matters-Certified-Braided-Thunderbolt-Charging-black/dp/B0BNW5C82D).

The issue is the same with both cables though.

**Comment by mkopec** [comment]
Might have been fixed by https://github.com/Dasharo/coreboot/pull/612 , need to retest on the latest code

**Comment by PerAstraAdDeum** [comment]
> Might have been fixed by [Dasharo/coreboot#612](https://github.com/Dasharo/coreboot/pull/612) , need to retest on the latest code

That would indeed be great! I'd be happy to test a beta release if there's one.

**Comment by PerAstraAdDeum** [comment]
Also, while talking about voltage:

![Image](https://github.com/user-attachments/assets/85b5ac3b-16c7-45a2-9c49-823b3ad112bf)

The V54's iGPU is constantly throttling, apparently because of reaching the voltage limit. Sadly I don't know what voltage the iGPU is currently drawing, even though that should be implemented in MangoHud already:

![Image](https://github.com/user-attachments/assets/14201187-85fe-41a3-be03-087304d99931)

I assume this is because of missing https://github.com/Dasharo/dasharo-issues/issues/820 ?

**Comment by SebastianCzapla** [comment]
When gathering reference performance values for this issue, I had trouble to properly benchmark with the Echo 11 dock.
This is somewhat related to https://github.com/Dasharo/dasharo-issues/issues/1081, but manifests across OSes in exactly opposite conditions.

Initial benchmarking of v54tu without and with dock were similar but inconclusive, however connecting the USB-C to HDMI display makes the benchmark barely respond, often crashing whole system (Ubuntu 24.10 with 6.11.0-19 generic kernel).
This seems suspicious in context of what happens on the Windows system: where USB ports and Ethernet connection of the dock randomly drops *unless* the USB-C to HDMI Display is connected.

I will recreate the setup again, and try gather dmesg logs if the machine is responsive at all.

**Comment by mkopec** [comment]
Voltage / current throttling should be fixed as of https://github.com/Dasharo/coreboot/pull/612

**Comment by PerAstraAdDeum** [comment]
> Voltage / current throttling should be fixed as of [Dasharo/coreboot#612](https://github.com/Dasharo/coreboot/pull/612)
> 
> > Sadly I don't know what voltage the iGPU is currently drawing
> 
> You can check detailed voltages and throttle reasons in HWInfo64

Happy to hear! I don't know exactly what the linked issue fixes though?
Also, I'm not on Windows, so HWInfo64 isn't available. Is there any way of checking the voltage in Linux?

**Comment by mkopec** [comment]
>  I don't know exactly what the linked issue fixes though?

That PR was initially meant to fix some other throttling issues on -TU models, in that case I narrowed it down to current / voltage throttling. If this is the same throttling then the issue might be fixed in the next release for this platform.

> Is there any way of checking the voltage in Linux?

I'm not aware of it sadly

**Comment by SebastianCzapla** [comment]
[dmesg-unigine-superposition-at-817.log](https://github.com/user-attachments/files/19448904/dmesg-unigine-superposition-at-817.log)

Here is a dmesg while testing this issue with a Echo 11 Dock.
The issue at ~817s in, happens when even slightest movement happens to the window of benchmark. This does not happen at all without the dock attached.
```
[  817.723135] workqueue: delayed_fput hogged CPU for >10000us 4 times, consider switching to WQ_UNBOUND
[  831.332198] workqueue: delayed_fput hogged CPU for >10000us 5 times, consider switching to WQ_UNBOUND
[  844.034255] workqueue: delayed_fput hogged CPU for >10000us 7 times, consider switching to WQ_UNBOUND
```
Disconnecting dock makes the issue go away immediately. Could this be related to lost performance? It definitely is a weird issue only specific to this dock.

**Comment by matmacieje** [comment]
Tested with INSYDE FIRMWARE on V540TU & V560TNE;

Windows 11: no reproduction
Ubuntu 24:

- V540TU: issue exists, but no specific traces in dmesg
- V560TNE: issue exists:
```
[ 1338.764924] workqueue: delayed_fput hogged CPU for >10000us 7 times, consider switching to WQ_UNBOUND
[ 1340.320926] workqueue: delayed_fput hogged CPU for >10000us 11 times, consider switching to WQ_UNBOUND
```

[v540tu _dmesg.txt](https://github.com/user-attachments/files/19630623/v540tu._dmesg.txt)

[v560tne_dmesg_hog.txt](https://github.com/user-attachments/files/19630626/v560tne_dmesg_hog.txt)

**Comment by matmacieje** [comment]
hogged CPU reproduction on V560TNE:

![Image](https://github.com/user-attachments/assets/bf584a01-1f37-4482-8ae0-68158de94863)

**Comment by matmacieje** [comment]
Execution results on V560TNE, Dasharo 1.0.0-rc1:

![Image](https://github.com/user-attachments/assets/6e92b488-68c1-42fa-9cc5-091c11e71b30)

**Comment by matmacieje** [comment]
hogged CPU reproduction on V540TU:

![Image](https://github.com/user-attachments/assets/61c0b949-bffd-4f0c-a98b-6384e5697dbf)

**Comment by mkopec** [comment]
I guess taking screenshots on GNOME Wayland is broken, NVIDIA probably uses X11 by default

**Comment by matmacieje** [comment]
Execution results on V540TU, Dasharo 1.0.0-rc1:

![Image](https://github.com/user-attachments/assets/78689f74-1a41-4851-b13a-9a603115d068)

**Comment by matmacieje** [comment]
> [@matmacieje](https://github.com/matmacieje) just making sure, you did update the EC as well, right?

@mkopec yes, 2025_04_08

**Comment by matmacieje** [comment]
Issue exists in Dasharo rc2, reproduced on Ubuntu 24, V560TNE.

[v560tne_dmesg_hog_rc2.txt](https://github.com/user-attachments/files/19667047/v560tne_dmesg_hog_rc2.txt)

**Comment by miczyg1** [comment]
I couldn't reproduce it on iGPU variant, results were comparable: https://github.com/Dasharo/dasharo-issues/issues/1320#issuecomment-2859202696

**Comment by PerAstraAdDeum** [comment]
> I couldn't reproduce it on iGPU variant, results were comparable: [#1320 (comment)](https://github.com/Dasharo/dasharo-issues/issues/1320#issuecomment-2859202696)

First Benchmark while on DC power, today at 14:44:

![Image](https://github.com/user-attachments/assets/fd331337-9a0f-4e55-95bd-df537780b7dc)

Pull DC connector, plug in TB4-connector. Second Benchmark, today at 14:48:

![Image](https://github.com/user-attachments/assets/9bbca9dc-6ab0-4073-b013-4cbabe26b82a)

I'm sorry. I can imagine you're all tired of this issue at this point. But I can assure it, is is very much still present on v1.0.0-rc3 for me. @miczyg1, what version were you running before?

**Comment by miczyg1** [comment]
> I'm sorry. I can imagine you're all tired of this issue at this point. But I can assure it, is is very much still present on v1.0.0-rc3 for me. [@miczyg1](https://github.com/miczyg1), what version were you running before?

Same, v1.0.0-rc3

**Comment by philipanda** [comment]
TL;DR:
No reproduction, not even once. There must be some missing detail we are not aware of.

Getting 4500-5000 scores on AC, on USB-PD with both the Wavlink UMD05 Pro (Rev.e) and with the Echo 11 Thunderbolt 4 Sonnettech, and on battery. Tested on Ubuntu 24.04 kernel 7.0.0 driver i915,    
Fedora 42 kernel 6.19.14 driver i915 and Windows 25H2 on DX11. I've used two stock NovaCustom 32GB RAM modules (W-NM56S532G), the CPU was Ultra 7 155H.
Every docking station and OS tested on:
- current `dasharo` branch, `35b6b8485419`
- v1.0.1
- v1.0.0
- v1.0.0-rc2 


When the USB-PD charger amperage reduction is removed (one of the changes from https://github.com/Dasharo/coreboot/pull/612 was reducing it, which doesn't seem to have an effect as the final amperage is quantized and 85 -> 93 fit in the same bucker) and I've increased the tcc_offset so that thermal throttling is not a thing, it gets us somewhere about 400 benchmark points, but that does not affect the issue.

It's important to use two RAM modules to achieve the 5000'ish scores reported by @PerAstraAdDeum, other runs form us seem to have used only one ram module which cuts the memory bandwidth available to the GPU significantly thus reducing the scores from ~4700-5000 to ~3300-3600 and the average FPS from ~36 to ~25.
But that is a global performance reduction. It does not cause any difference between PD and AC.

Two smaller things found along the way. The CPU was throttling on temperature at 80C (on Windows that's even at idle). Raising  
that to 100C gained about 400 points. And the EC caps USB-PD charging current at 85%, so USB-PD registered  
as an 80W supply while the barrel jack registered as 90W. Removing the cap makes them identical but changed 
no benchmark result.

Checked:
 - CPU and platform power limits
    <details>

     What coreboot "thinks" it set:
     ```console
     sudo cbmem -c | grep -iE "PL1|PL2|PL4|TDP|power limits"
     [INFO ]  CPU TDP = 28 Watts
     [INFO ]  CPU PL1 = 35 Watts
     [INFO ]  CPU PL2 = 64 Watts
     [INFO ]  CPU PL4 = 90 Watts
     [DEBUG]  Configured power limits for SA PCI ID: 0x7d01
     ```
     ```console
     $ echo "PL1: $(( $(sudo rdmsr -p 0 -f 14:0  -d 0x610) / 8 )) W"
     PL1: 35 W
     $ echo "PL2: $(( $(sudo rdmsr -p 0 -f 46:32 -d 0x610) / 8 )) W"
     PL2: 64 W
     $ echo "PL4: $(( $(sudo rdmsr -p 0 -f 14:0 -d 0x601) / 8 )) W"
     PL4: 90W
     ```

     The two sources should match.
    </details>
    PL1 is generally not achieved for long, so they're good.

- PD contracts with the chargers, current limits
    <details>

    We can get some information on that from an `ec_sys` module, or form a debug EC build. 
    The Fedora kernel does not have the support for that, but Ubuntu does. I'm not sure about other distros.
    ```
    $ sudo modprobe ec_sys
    $ echo "$(sudo dd if=/sys/kernel/debug/ec/ec0/io bs=1 skip=213 count=4 status=none | od -An -tu4) mW"
    ```
    For the debug EC path:
    Flashing the EC is a bit dangerous as recovery is difficult.
    From [this doc](https://docs.dasharo.com/kb/laptop-dev-env/#debugging) we only need to change the debug level to 4 and rebuild.
    Then, still in the `ec` repository, we can build the `dasharo_ectool` to read the logs from a running system.
    you might need to install the rust toolchain for that.
    ```
    ~/ec/ $ make BOARD=novacustom/v540tu console_internal
    ```
    The interesting lines printed by the EC with debug logging are:
    ```
    90 W supply detected
    CHARGER LIMIT 4740 mA
    AdapterCurrent: 0480
    ```
    The exact values might differ, but can give an idea about what power limit was negotiated with the charger.
    </details>
- Thermal throttling
    <details>

     It is a thing, with the conservative 30 C tcc_offset we've got as a default the CPU throttles at just 80 C, which is constantly, even when idle, on Windows, and very often, but rather under a load, on Linux.
    Increasing the throttle temperature to 100 C helps with the performance just a bit, couple hunrded points at most, unrelated to the issue.
   </details>
- Miscellaneous firwmare changes tried that did not change anything observable
    <details>

    - Disabling the PsysPLs in EC `-DPOWER_LIMIT_PSYS_ENABLE=0` in makefile
    - Setting CepEnable and EnableFastVmode in FSP to 0 to disable overcurrent protections in the CPU which could've been causing some throttling
    </details>
- Throttling reasons
    <details>

    - ThrottleStop (Windows) nothing interesting there, constant thermal throttling and EDP OTHER during full load. PL1 and PL2 are sometimes triggered under load, but not constantly.
    - a Linux alternative is exposed by the kernel directly:
    ```
    # grep -H . /sys/class/drm/card*/gt/gt0/throttle_reason_*
    /sys/class/drm/card1/gt/gt0/throttle_reason_pl1:0
    /sys/class/drm/card1/gt/gt0/throttle_reason_pl2:0
    /sys/class/drm/card1/gt/gt0/throttle_reason_pl4:1
    /sys/class/drm/card1/gt/gt0/throttle_reason_prochot:0
    /sys/class/drm/card1/gt/gt0/throttle_reason_ratl:0
    /sys/class/drm/card1/gt/gt0/throttle_reason_status:1
    /sys/class/drm/card1/gt/gt0/throttle_reason_thermal:0
    /sys/class/drm/card1/gt/gt0/throttle_reason_vr_tdc:0
    /sys/class/drm/card1/gt/gt0/throttle_reason_vr_thermalert:0
    ```
    ^ that's from a Ubuntu run on USB-PD from an echo 11 tb4 dock. It got `4550` score, so everything was fine.

    </details>
- Some screenshots from the benchmark results. I did not write down every execution as the results were always in ~200 points range and didn't seem interesting.
    <details>
    Fedora AC
    <img width="1918" height="1098" alt="Image" src="https://github.com/user-attachments/assets/5ad41b8c-cc7a-49be-b1e9-ce112a44cdd5" />
    Fedora USB-PD
    <img width="1918" height="1098" alt="Image" src="https://github.com/user-attachments/assets/f75b84ab-82d8-491f-9b35-2d63d073b16a" />
    Ubuntu AC
    <img width="1842" height="1163" alt="Image" src="https://github.com/user-attachments/assets/0cce8bdf-a3e2-4845-b617-4bfecc68a0bb" />
    Ubuntu USB-PD
    <img width="1842" height="1163" alt="Image" src="https://github.com/user-attachments/assets/daff4424-1d01-43df-8ae2-fe3110c3ab35" />
    Windows AC
    <img width="2048" height="1280" alt="Image" src="https://github.com/user-attachments/assets/b28d8f30-6854-4548-a13f-1cd73660faad" />
    Windows USB-PD
    <img width="2048" height="1280" alt="Image" src="https://github.com/user-attachments/assets/01931a55-d7db-47a9-b9d8-b2f3dc8f7afa" />

    </details>
It would help to get the outputs of the commands described in the expandable `▶ Details`.
